### PR TITLE
Chunk streaming improvements

### PR DESCRIPTION
**Pull request by tigerw**

### CODE REVIEW

**Comment by 12xx12** [comment]
Is this the fix for invisible mobs?

**Comment by tigerw** [comment]
For teleporting to a faraway land and back again breaking chunk sending. Hopefully fixes #4531. I'd also like to work on the other chunk related issues #5245 and #3696.

**Comment by tigerw** [comment]
Copied from https://github.com/cuberite/cuberite/issues/5245#issuecomment-867948533

> Having fully loaded the chunks in your area, when teleporting to a faraway location and immediately back, the server doesn't send chunks anymore. This is because after chunks are fully loaded, an internal value is set to your current chunk to stop further chunk streaming. Teleporting away and back unloads every chunk clientside, but because that internal value is unchanged, chunk streaming remains stopped.

To fix this m_LastSentChunk is reset to a bogus value whenever the player's current chunk doesn't match.

**Comment by 12xx12** [comment]
The protoproxy tool is broken on compile - I really don't know if this good working with the broken API

**Comment by 12xx12** [comment]
I will take a look at this

File: src/ClientHandle.cpp [hunk 1]
```diff
@@ -388,25 +392,30 @@ void cClientHandle::FinishAuthenticate(const AString & a_Name, const cUUID & a_U
...
 	if ((m_LastStreamedChunkX == ChunkPosX) && (m_LastStreamedChunkZ == ChunkPosZ))
 	{
 		// All chunks are already loaded. Abort loading.
-		return true;
+		return;
+	}
+	else
+	{
+		m_LastStreamedChunkX = 0x7fffffff;
+		m_LastStreamedChunkZ = 0x7fffffff;
```
Comment (12xx12) [inline src/ClientHandle.cpp:410]: Why those magic numbers?

File: src/mbedTLS++/AesCfb128Encryptor.cpp [hunk 2]
```diff
@@ -43,8 +43,8 @@ void cAesCfb128Encryptor::Init(const Byte a_Key[16], const Byte a_IV[16])
 
 
 
-void cAesCfb128Encryptor::ProcessData(std::byte * const a_EncryptedOut, const std::byte * const a_PlainIn, size_t a_Length)
+void cAesCfb128Encryptor::ProcessData(std::byte * const a_PlainIn, const size_t a_Length)
 {
 	ASSERT(IsValid());  // Must Init() first
-	mbedtls_aes_crypt_cfb8(&m_Aes, MBEDTLS_AES_ENCRYPT, a_Length, m_IV, reinterpret_cast<const unsigned char *>(a_PlainIn), reinterpret_cast<unsigned char *>(a_EncryptedOut));
+	mbedtls_aes_crypt_cfb8(&m_Aes, MBEDTLS_AES_ENCRYPT, a_Length, m_IV, reinterpret_cast<const unsigned char *>(a_PlainIn), reinterpret_cast<unsigned char *>(a_PlainIn));
```
Comment (12xx12) [inline src/mbedTLS++/AesCfb128Encryptor.cpp:49]: Now we do inplace encryption?

**Review by 12xx12 — COMMENTED** [review]
So to summarize:
- we now encrypt sent data in place
- you rephrased some code that triggers the chunk sending
right?
- we now can process the data sent by a protocol before sending (e.g encrypt) instead of hiding this somewhere (I like this)

File: src/mbedTLS++/AesCfb128Encryptor.cpp (hunk 2 above)
Comment (tigerw) [reply src/mbedTLS++/AesCfb128Encryptor.cpp:49]: Yes, on the protocol out buffer.

**Comment by tigerw** [comment]
Yes. For the third, if you mean additional processing by plugins, in theory but it's binary data so not too useful.

**Comment by 12xx12** [comment]
Or disabling decryption for testing purposes - the hardcore way

File: src/ClientHandle.cpp (hunk 1 above)
Comment (12xx12) [reply src/ClientHandle.cpp:410]: I see the sense of those values, but an explanation would be nice. or a `#define INSANEVALUE 0x7fffffff`

Comment (12xx12) [reply src/ClientHandle.cpp:410]: Now I see the big change - you invalidate the last sent chunk if we're not in it 💡

**Review by 12xx12 — APPROVED** [review]

**Comment by 12xx12** [comment]
Note that some compiling is still broken

**Comment by tigerw** [comment]
Thanks, let me change the constant.

**Comment by 12xx12** [comment]
Looks good to me

**Comment by 12xx12** [comment]
Is there a reason you don't squash your PRs?

I think it's nicer to have a commit for each PR and have a number attached to it to check the thoughts by the developer.

In the end, it's just convenience. So it's not that important

**Comment by Janmm14** [comment]
you could use a merge commit with the PR number

**Comment by 12xx12** [comment]
I usually just push the big green button and GitHub does everything else :D

**Comment by tigerw** [comment]
Yeah you're right, this could've been squashed

**Comment by 12xx12** [comment]
There is always a next PR to merge 😄

**Pull request merged**